File: php/src/Metrics/HdrHistogram.php
<?php

declare(strict_types=1);

namespace RespBench\Metrics;

/**
 * Minimal pure-PHP HdrHistogram compatible with the HdrHistogram bucketing scheme
 * used by the Java/Ruby engines (lowest=1, highest=600_000_000, sigfig=3).
 *
 * Implements the standard HdrHistogram value→index mapping so that the encoded
 * V2 payload (see HdrEncoder) is byte-compatible with Java's
 * Histogram.encodeIntoCompressedByteBuffer().
 *
 * Reference: https://github.com/HdrHistogram/HdrHistogram
 */
final class HdrHistogram
{
    public readonly int $lowestTrackableValue;
    public readonly int $highestTrackableValue;
    public readonly int $significantFigures;

    private readonly int $subBucketHalfCountMagnitude;
    private readonly int $subBucketHalfCount;
    private readonly int $subBucketCount;
    private readonly int $subBucketMask;
    private readonly int $unitMagnitude;
    private readonly int $bucketCount;
    public readonly int $countsLen;

    /** @var array<int,int> sparse index => count */
    private array $counts = [];

    private int $totalCount = 0;
    private int $minNonZeroValue = PHP_INT_MAX;
    private int $maxValue = 0;

    public function __construct(
        int $lowestTrackableValue = 1,
        int $highestTrackableValue = 600_000_000,
        int $significantFigures = 3,
    ) {
        $this->lowestTrackableValue = $lowestTrackableValue;
        $this->highestTrackableValue = $highestTrackableValue;
        $this->significantFigures = $significantFigures;

        $largestValueWithSingleUnitResolution = 2 * (int) (10 ** $significantFigures);
        $subBucketCountMagnitude = (int) ceil(log($largestValueWithSingleUnitResolution, 2));
        $this->subBucketHalfCountMagnitude = max($subBucketCountMagnitude, 1) - 1;

        $this->unitMagnitude = (int) floor(log($lowestTrackableValue, 2));

        $this->subBucketCount = 2 ** ($this->subBucketHalfCountMagnitude + 1);
        $this->subBucketHalfCount = $this->subBucketCount >> 1;
        $this->subBucketMask = ($this->subBucketCount - 1) << $this->unitMagnitude;

        // Number of buckets needed to cover the highest trackable value.
        $smallestUntrackableValue = $this->subBucketCount << $this->unitMagnitude;
        $bucketsNeeded = 1;
        while ($smallestUntrackableValue < $highestTrackableValue) {
            if ($smallestUntrackableValue > (PHP_INT_MAX >> 1)) {
                $bucketsNeeded++;
                break;
            }
            $smallestUntrackableValue <<= 1;
            $bucketsNeeded++;
        }
        $this->bucketCount = $bucketsNeeded;
        $this->countsLen = ($this->bucketCount + 1) * ($this->subBucketCount >> 1);
    }

    public function record(int $value): void
    {
        if ($value < 0) {
            return;
        }
        $index = $this->countsIndexFor($value);
        $this->counts[$index] = ($this->counts[$index] ?? 0) + 1;
        $this->totalCount++;

        if ($value > $this->maxValue) {
            $this->maxValue = $value;
        }
        if ($value !== 0 && $value < $this->minNonZeroValue) {
            $this->minNonZeroValue = $value;
        }
    }

    public function recordValueWithCount(int $value, int $count): void
    {
        if ($value < 0 || $count <= 0) {
            return;
        }
        $index = $this->countsIndexFor($value);
        $this->counts[$index] = ($this->counts[$index] ?? 0) + $count;
        $this->totalCount += $count;

        if ($value > $this->maxValue) {
            $this->maxValue = $value;
        }
        if ($value !== 0 && $value < $this->minNonZeroValue) {
            $this->minNonZeroValue = $value;
        }
    }

    public function merge(self $other): void
    {
        for ($i = 0; $i < $other->countsLen; $i++) {
            $count = $other->rawCountAt($i);
            if ($count > 0) {
                $value = $other->valueFromIndex($i);
                $this->recordValueWithCount($value, $count);
            }
        }
    }

    public function totalCount(): int
    {
        return $this->totalCount;
    }

    public function min(): int
    {
        if ($this->totalCount === 0) {
            return 0;
        }

        // If only zero-valued samples were recorded, min is 0.
        return $this->minNonZeroValue === PHP_INT_MAX ? 0 : $this->minNonZeroValue;
    }

    public function max(): int
    {
        return $this->maxValue;
    }

    public function valueAtPercentile(float $percentile): int
    {
        if ($this->totalCount === 0) {
            return 0;
        }

        $requestedPercentile = min(max($percentile, 0.0), 100.0);
        $countAtPercentile = (int) ceil(($requestedPercentile / 100.0) * $this->totalCount);
        $countAtPercentile = max($countAtPercentile, 1);

        $total = 0;
        for ($i = 0; $i < $this->countsLen; $i++) {
            $total += $this->rawCountAt($i);
            if ($total >= $countAtPercentile) {
                $valueAtIndex = $this->valueFromIndex($i);

                return $this->highestEquivalentValue($valueAtIndex);
            }
        }

        return $this->maxValue;
    }

    public function rawCountAt(int $index): int
    {
        return $this->counts[$index] ?? 0;
    }

    /**
     * Highest index with a non-zero count, plus one (the "relevant length").
     */
    public function relevantLength(): int
    {
        if ($this->counts === []) {
            return 0;
        }

        return max(array_keys($this->counts)) + 1;
    }

    // --- HdrHistogram index math (mirrors the Java implementation) ---

    private function countsIndexFor(int $value): int
    {
        $bucketIndex = $this->bucketIndexFor($value);
        $subBucketIndex = $this->subBucketIndexFor($value, $bucketIndex);

        return $this->countsIndex($bucketIndex, $subBucketIndex);
    }

    private function bucketIndexFor(int $value): int
    {
        $pow2ceiling = self::bitLength($value | $this->subBucketMask);

        return $pow2ceiling - $this->unitMagnitude - ($this->subBucketHalfCountMagnitude + 1);
    }

    private function subBucketIndexFor(int $value, int $bucketIndex): int
    {
        return $value >> ($bucketIndex + $this->unitMagnitude);
    }

    private function countsIndex(int $bucketIndex, int $subBucketIndex): int
    {
        $bucketBaseIndex = ($bucketIndex + 1) << $this->subBucketHalfCountMagnitude;
        $offsetInBucket = $subBucketIndex - $this->subBucketHalfCount;

        return $bucketBaseIndex + $offsetInBucket;
    }

    public function valueFromIndex(int $index): int
    {
        $bucketIndex = ($index >> $this->subBucketHalfCountMagnitude) - 1;
        $subBucketIndex = ($index & ($this->subBucketHalfCount - 1)) + $this->subBucketHalfCount;

        if ($bucketIndex < 0) {
            $subBucketIndex -= $this->subBucketHalfCount;
            $bucketIndex = 0;
        }

        return $subBucketIndex << ($bucketIndex + $this->unitMagnitude);
    }

    private function highestEquivalentValue(int $value): int
    {
        return $this->nextNonEquivalentValue($value) - 1;
    }

    private function nextNonEquivalentValue(int $value): int
    {
        return $this->lowestEquivalentValue($value) + $this->sizeOfEquivalentValueRange($value);
    }

    private function lowestEquivalentValue(int $value): int
    {
        $bucketIndex = $this->bucketIndexFor($value);
        $subBucketIndex = $this->subBucketIndexFor($value, $bucketIndex);

        return $subBucketIndex << ($bucketIndex + $this->unitMagnitude);
    }

    private function sizeOfEquivalentValueRange(int $value): int
    {
        $bucketIndex = $this->bucketIndexFor($value);

        return 1 << ($this->unitMagnitude + $bucketIndex);
    }

    private static function bitLength(int $value): int
    {
        $length = 0;
        while ($value > 0) {
            $value >>= 1;
            $length++;
        }

        return $length;
    }
}

File: go/internal/metrics/hdrhistogram.go
// Package metrics implements latency collection, HdrHistogram encoding, and
// NDJSON output compatible with the reference engines.
package metrics

import "math"

// HdrHistogram is a minimal pure-Go HdrHistogram implementing the standard
// value→index bucketing (lowest=1, highest=600_000_000, sigfig=3) so the encoded
// V2 payload (see HdrEncoder) is byte-compatible with Java's
// Histogram.encodeIntoCompressedByteBuffer().
//
// Reference: https://github.com/HdrHistogram/HdrHistogram
type HdrHistogram struct {
	LowestTrackableValue  int64
	HighestTrackableValue int64
	SignificantFigures    int

	subBucketHalfCountMagnitude int
	subBucketHalfCount          int64
	subBucketCount              int64
	subBucketMask               int64
	unitMagnitude               int
	bucketCount                 int
	CountsLen                   int

	counts   map[int]int64 // sparse index → count
	total    int64
	minValue int64
	maxValue int64
}

// NewHdrHistogram builds a histogram with the standard resp-bench parameters.
func NewHdrHistogram(lowest, highest int64, sigFigs int) *HdrHistogram {
	h := &HdrHistogram{
		LowestTrackableValue:  lowest,
		HighestTrackableValue: highest,
		SignificantFigures:    sigFigs,
		counts:                map[int]int64{},
		minValue:              math.MaxInt64,
	}

	largestValueWithSingleUnitResolution := int64(2 * int(math.Pow(10, float64(sigFigs))))
	subBucketCountMagnitude := int(math.Ceil(math.Log2(float64(largestValueWithSingleUnitResolution))))
	if subBucketCountMagnitude < 1 {
		subBucketCountMagnitude = 1
	}
	h.subBucketHalfCountMagnitude = subBucketCountMagnitude - 1

	h.unitMagnitude = int(math.Floor(math.Log2(float64(lowest))))

	h.subBucketCount = int64(1) << (h.subBucketHalfCountMagnitude + 1)
	h.subBucketHalfCount = h.subBucketCount >> 1
	h.subBucketMask = (h.subBucketCount - 1) << h.unitMagnitude

	smallestUntrackable := h.subBucketCount << h.unitMagnitude
	bucketsNeeded := 1
	for smallestUntrackable < highest {
		if smallestUntrackable > (math.MaxInt64 >> 1) {
			bucketsNeeded++
			break
		}
		smallestUntrackable <<= 1
		bucketsNeeded++
	}
	h.bucketCount = bucketsNeeded
	h.CountsLen = (h.bucketCount + 1) * int(h.subBucketCount>>1)
	return h
}

// NewDefaultHdrHistogram returns the shared 1µs..600s / 3-sigfig histogram used
// by every engine. Using a different max would make the encoding incompatible.
func NewDefaultHdrHistogram() *HdrHistogram {
	return NewHdrHistogram(1, 600_000_000, 3)
}

// Record adds a single observation.
func (h *HdrHistogram) Record(value int64) { h.RecordValueWithCount(value, 1) }

// RecordValueWithCount adds count observations of value.
func (h *HdrHistogram) RecordValueWithCount(value, count int64) {
	if value < 0 || count <= 0 {
		return
	}
	idx := h.countsIndexFor(value)
	h.counts[idx] += count
	h.total += count
	if value > h.maxValue {
		h.maxValue = value
	}
	if value != 0 && value < h.minValue {
		h.minValue = value
	}
}

// Merge folds another histogram into this one by re-recording at bucket floors.
func (h *HdrHistogram) Merge(other *HdrHistogram) {
	for i := 0; i < other.CountsLen; i++ {
		if c := other.RawCountAt(i); c > 0 {
			h.RecordValueWithCount(other.ValueFromIndex(i), c)
		}
	}
}

// TotalCount returns the number of recorded observations.
func (h *HdrHistogram) TotalCount() int64 { return h.total }

// Min returns Java getMinValue()-equivalent: the lowest populated bucket's value.
func (h *HdrHistogram) Min() int64 {
	if h.total == 0 {
		return 0
	}
	return h.ValueAtPercentile(0)
}

// Max returns Java getMaxValue()-equivalent: the top populated bucket's ceiling.
func (h *HdrHistogram) Max() int64 {
	if h.total == 0 {
		return 0
	}
	return h.ValueAtPercentile(100)
}

// ValueAtPercentile returns the value at the given percentile [0,100].
func (h *HdrHistogram) ValueAtPercentile(p float64) int64 {
	if h.total == 0 {
		return 0
	}
	if p < 0 {
		p = 0
	}
	if p > 100 {
		p = 100
	}
	countAt := int64(math.Ceil((p / 100.0) * float64(h.total)))
	if countAt < 1 {
		countAt = 1
	}
	var running int64
	for i := 0; i < h.CountsLen; i++ {
		running += h.RawCountAt(i)
		if running >= countAt {
			return h.highestEquivalentValue(h.ValueFromIndex(i))
		}
	}
	return h.maxValue
}

// RawCountAt returns the count at a counts-array index.
func (h *HdrHistogram) RawCountAt(index int) int64 { return h.counts[index] }

// RelevantLength returns the highest non-zero index + 1.
func (h *HdrHistogram) RelevantLength() int {
	maxIdx := -1
	for idx := range h.counts {
		if h.counts[idx] != 0 && idx > maxIdx {
			maxIdx = idx
		}
	}
	return maxIdx + 1
}

// --- HdrHistogram index math (mirrors the Java implementation) ---

func (h *HdrHistogram) countsIndexFor(value int64) int {
	bucketIndex := h.bucketIndexFor(value)
	subBucketIndex := h.subBucketIndexFor(value, bucketIndex)
	return h.countsIndex(bucketIndex, subBucketIndex)
}

func (h *HdrHistogram) bucketIndexFor(value int64) int {
	pow2ceiling := bitLength(value | h.subBucketMask)
	return pow2ceiling - h.unitMagnitude - (h.subBucketHalfCountMagnitude + 1)
}

func (h *HdrHistogram) subBucketIndexFor(value int64, bucketIndex int) int64 {
	return value >> uint(bucketIndex+h.unitMagnitude)
}

func (h *HdrHistogram) countsIndex(bucketIndex int, subBucketIndex int64) int {
	bucketBaseIndex := (bucketIndex + 1) << h.subBucketHalfCountMagnitude
	offsetInBucket := int(subBucketIndex - h.subBucketHalfCount)
	return bucketBaseIndex + offsetInBucket
}

// ValueFromIndex reconstructs the bucket floor value for a counts-array index.
func (h *HdrHistogram) ValueFromIndex(index int) int64 {
	bucketIndex := (index >> h.subBucketHalfCountMagnitude) - 1
	subBucketIndex := int64(index&int(h.subBucketHalfCount-1)) + h.subBucketHalfCount
	if bucketIndex < 0 {
		subBucketIndex -= h.subBucketHalfCount
		bucketIndex = 0
	}
	return subBucketIndex << uint(bucketIndex+h.unitMagnitude)
}

func (h *HdrHistogram) highestEquivalentValue(value int64) int64 {
	return h.nextNonEquivalentValue(value) - 1
}

func (h *HdrHistogram) nextNonEquivalentValue(value int64) int64 {
	return h.lowestEquivalentValue(value) + h.sizeOfEquivalentValueRange(value)
}

func (h *HdrHistogram) lowestEquivalentValue(value int64) int64 {
	bucketIndex := h.bucketIndexFor(value)
	subBucketIndex := h.subBucketIndexFor(value, bucketIndex)
	return subBucketIndex << uint(bucketIndex+h.unitMagnitude)
}

func (h *HdrHistogram) sizeOfEquivalentValueRange(value int64) int64 {
	bucketIndex := h.bucketIndexFor(value)
	return int64(1) << uint(h.unitMagnitude+bucketIndex)
}

// bitLength returns the number of bits needed to represent value (>=0).
func bitLength(value int64) int {
	length := 0
	for value > 0 {
		value >>= 1
		length++
	}
	return length
}
